kroxylicious/src/main/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilter.java

```diff
@@ -0,0 +1,224 @@
+import org.apache.kafka.common.message.MetadataRequestData;
+import org.apache.kafka.common.message.MetadataResponseData;
+import org.apache.kafka.common.message.OffsetCommitRequestData;
+import org.apache.kafka.common.message.OffsetCommitResponseData;
+import org.apache.kafka.common.message.OffsetFetchRequestData;
+import org.apache.kafka.common.message.OffsetFetchResponseData;
+import org.apache.kafka.common.message.OffsetForLeaderEpochRequestData;
+import org.apache.kafka.common.message.OffsetForLeaderEpochResponseData;
+import org.apache.kafka.common.message.ProduceRequestData;
+import org.apache.kafka.common.message.ProduceResponseData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.filter.CreateTopicsRequestFilter;
+import io.kroxylicious.proxy.filter.CreateTopicsResponseFilter;
+import io.kroxylicious.proxy.filter.DeleteTopicsRequestFilter;
+import io.kroxylicious.proxy.filter.DeleteTopicsResponseFilter;
+import io.kroxylicious.proxy.filter.FetchRequestFilter;
+import io.kroxylicious.proxy.filter.FetchResponseFilter;
+import io.kroxylicious.proxy.filter.KrpcFilterContext;
+import io.kroxylicious.proxy.filter.ListOffsetsRequestFilter;
+import io.kroxylicious.proxy.filter.ListOffsetsResponseFilter;
+import io.kroxylicious.proxy.filter.MetadataRequestFilter;
+import io.kroxylicious.proxy.filter.MetadataResponseFilter;
+import io.kroxylicious.proxy.filter.OffsetCommitRequestFilter;
+import io.kroxylicious.proxy.filter.OffsetCommitResponseFilter;
+import io.kroxylicious.proxy.filter.OffsetFetchRequestFilter;
+import io.kroxylicious.proxy.filter.OffsetFetchResponseFilter;
+import io.kroxylicious.proxy.filter.OffsetForLeaderEpochRequestFilter;
+import io.kroxylicious.proxy.filter.OffsetForLeaderEpochResponseFilter;
+import io.kroxylicious.proxy.filter.ProduceRequestFilter;
+import io.kroxylicious.proxy.filter.ProduceResponseFilter;
+import io.kroxylicious.proxy.internal.filter.FilterConfig;
+
+/**
+ * Simple multi-tenant filter.
+ *
+ * Uses the first component of a fully-qualified host name as a tenant prefix.
+ * This tenant prefix is prepended to the kafka resources name in order to present an isolated
+ * environment for each tenant.
+ *
+ * TODO prefix other resources e.g. group names, transaction ids
+ * TODO disallow the use of topic uids belonging to one tenant by another.
+ */
+public class MultiTenantTransformationFilter
+        implements CreateTopicsRequestFilter, CreateTopicsResponseFilter,
+        DeleteTopicsRequestFilter, DeleteTopicsResponseFilter,
+        MetadataRequestFilter, MetadataResponseFilter,
+        ProduceRequestFilter, ProduceResponseFilter,
+        ListOffsetsRequestFilter, ListOffsetsResponseFilter,
+        FetchRequestFilter, FetchResponseFilter,
+        OffsetFetchRequestFilter, OffsetFetchResponseFilter,
+        OffsetCommitRequestFilter, OffsetCommitResponseFilter,
+        OffsetForLeaderEpochRequestFilter, OffsetForLeaderEpochResponseFilter {
```

Comment:
It _might_ be worth adding an `interface` (to the public protocol filter API package) which extends all the filters of requests and response which contain topic names. Likewise group ids, transactional ids etc. That way this filter would only be extending those three interfaces. The benefit would be that it's a single place to update if we had other filters which wanted to intercept all topic names.

Reply:
Yes, maybe that would help.  How would you classify RPCs such as `AlterConfigs` and the `CreateACL` where the payload might refer to a topic name so might need a topic name mapping?

I had been wondering about how unmaintainable a multi-tenant filter  implementation would be once it has implemented all the methods for topic, group and transaction mapping.  Maybe multi-tenant would better be comprise a series of filters each focused on a separate concern (topic/group/etc).    Not sure right now.  

I like the suggestion, but I think I'd like to reflect for a while.

Reply:
I don't see multi tenancy as being a one and done step, though I think this is close to making it one, by way of example we probably want to enforce partition limits on a per tenant basis thus any enforcement filter would need to be tenant aware. 

The other thing I'm wondering about is filter ordering, do we need some mechanism for filters to declare that they need the tenant to be determined before they appear in the chain? Probably not even in the medium term as long as the filter definition order is deterministic from the configuration but worth thinking about. One alternative might be to add a `TenancyAware` marker interface which would allow us to ensure they are in the filter chain behind the filter which determines the tenant. 

My initial model was that we would have one filter upfront which determines the tenant ID and thus everything downstream of that could take appropriate action.